src/main/java/org/sharedid/endpoint/handler/SetAuditCookieHandler.java

```diff
@@ -0,0 +1,135 @@
+
+    private Meter createAuditCookieMeter;
+    private Meter createAuditCookieFailedMeter;
+    private Meter renewAuditCookieMeter;
+    private Meter serializeAuditCookieFailedMeter;
+
+    @Autowired
+    public SetAuditCookieHandler(LocationService locationService,
+                                 AuditCookieService auditCookieService,
+                                 MetricRegistry metricRegistry,
+                                 @Value("${cookie.audit.name}") String auditCookieName,
+                                 @Value("${cookie.audit.ttl}") long auditCookieTtl,
+                                 @Value("${cookie.audit.cipher-key}") String cipherKey,
+                                 @Value("${cookies.secure}") Boolean isSecureCookiesEnabled) {
+        this.locationService = locationService;
+        this.auditCookieService = auditCookieService;
+        this.auditCookieName = auditCookieName;
+        this.auditCookieTtl = auditCookieTtl;
+        this.cipherKey = cipherKey;
+        this.isSecureCookiesEnabled = isSecureCookiesEnabled;
+
+        this.createAuditCookieMeter = metricRegistry.meter(METRIC_CREATE_AUDIT_COOKIE);
+        this.createAuditCookieFailedMeter = metricRegistry.meter(METRIC_CREATE_AUDIT_COOKIE_FAILED);
+        this.renewAuditCookieMeter = metricRegistry.meter(METRIC_RENEW_AUDIT_COOKIE);
+        this.serializeAuditCookieFailedMeter = metricRegistry.meter(METRIC_SERIALIZE_AUDIT_COOKIE_FAILED);
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+        DataContext dataContext = DataContext.from(routingContext);
+
+        AuditCookie auditCookie = dataContext.getAuditCookie();
+
+        Future<AuditCookie> auditCookieFuture;
+        if (auditCookie == null) {
+            logger.debug("Creating audit cookie");
+            createAuditCookieMeter.mark();
+            auditCookieFuture = getNewAuditCookie(dataContext, routingContext);
+        } else {
+            logger.debug("Renewing audit cookie");
+            renewAuditCookieMeter.mark();
+            auditCookie.setRenewedTimestampSeconds(Instant.now().toEpochMilli());
+            auditCookieFuture = Future.succeededFuture(auditCookie);
+        }
+
+        auditCookieFuture.onComplete(result -> {
+            if (result.failed()) {
+                logger.error("Failed to create an audit cookie", result.cause());
+                createAuditCookieFailedMeter.mark();
+                routingContext.next();
+                return;
+            }
+
+            try {
+                Cookie cookie = result.result().toCookie(auditCookieName, cipherKey, auditCookieTtl);
+
+                boolean isSecure;
+                if (isSecureCookiesEnabled != null) {
+                    isSecure = isSecureCookiesEnabled;
+                } else {
+                    isSecure = routingContext.request().scheme().equals("https");
+                }
+
+                cookie.setSecure(isSecure);
```

Comment:
It doesn't look like SameSite is set for audit cookies.

Reply:
My bad on this one.  Sorry I missed the call to ExtraHeadersCookie somehow.

Reply:
We should get rid of the ExtraHeadersCookie anyway. Primary development happened before vert.x added that field to the standard framework, so ExtraHeadersCookie is more of a work around.